File: C++/139. Word Break.cpp
class Solution {
public:
    bool wordBreak(string s, vector<string>& wordDict) {
        unordered_set<string> dict(wordDict.begin(),wordDict.end());
        int n=s.size();
        vector<bool>dp(n+1,0);
        dp[0]=1;
        for(int i=0;i<n;i++){
            if(!dp[i]){continue;}
            for(int j=i+1;j<n+1;j++){
                if(dict.count(s.substr(i,j-i)))
                dp[j]=1;
            }
        }
        return dp[n];
    }
};

File: C#/139. Word Break.cs
public class Solution {
    public bool WordBreak(string s, IList<string> wordDict) {
        var dict = wordDict.ToHashSet();
        int n=s.Length;
        bool[] dp = new bool[n+1];
        dp[0]=true;
        for(int i=0;i<n;i++){
            if(!dp[i]){continue;}
            for(int j=i+1;j<n+1;j++){
                if(dict.Contains(s.Substring(i,j-i)))
                dp[j]=true;
            }
        }
        return dp[n];
    }
}
